### PR TITLE
Update project charter following JCSDA/EMC/EPIC meeting 2023/07/19

### LINKED ISSUE #493: Update project charter
**Is your feature request related to a problem? Please describe.**
We need to update the project charter with everything we have agreed on in the past months:
- installers/maintainers on each platform
- installation destinations
- directory structure, build matrices
- procedures for rolling out, updating and removing existing spack-stack installations

**Additional context**
See also:
- https://github.com/NOAA-EMC/spack-stack/issues/90
- (access restricted) https://docs.google.com/spreadsheets/d/1j31C2WrG3hNAQAKdey8CbOx9eek_kGMsUVuHeBUhamg/edit#gid=532319951
- https://github.com/NOAA-EMC/spack-stack/blob/develop/project_charter.md

### PR DESCRIPTION
**Pull request by climbfuji**
### Summary

Update project charter with what was discussed on July 19 between JCSDA, NOAA-EMC, and EPIC. Please update/correct this as needed and merge when satisfied (@srherbener can approve for JCSDA).

You can render the markdown in any mark down live preview page.

### Testing

n/a

### Applications affected

n/a

### Systems affected

n/a

### Dependencies

n/a

### Issue(s) addressed

Closes #493 

### Checklist
- [x] This PR addresses one issue/problem/enhancement, or has a very good reason for not doing so.
- [ ] ~~These changes have been tested on the affected systems and applications.~~
- [ ] ~~All dependency PRs/issues have been resolved and this PR can be merged.~~

### CODE REVIEW

**Comment by climbfuji** [comment]
@arunchawla-NOAA @junwang-noaa @ytremolet FYI

File: project_charter.md [hunk 1]
```diff
@@ -103,3 +110,85 @@ Problems in spack package files will be solved by programmers of the
...
+
+Test installations are limited to one platform at a time (depending on the developer’s needs). The installers for that platform (see table in https://spack-stack.readthedocs.io/en/1.4.1/PreConfiguredSites.html) are responsible for installing it.
+
+Test installations are made outside the official spack-stack release environment using chained environments (if possible, otherwise completely new install trees), never in the official environments. This currently requires the developer to add an additional module use statement, as in the example below. The spack-stack developers are working on a solution to make the upstream modules available to the chained environment to make life easier for the users.
+
+```
+module use /path/to/separate/spack-stack-1.4.1-testing/envs/fms-2023.02-chained/install/modulefiles/openmpi/4.1.6/gcc/11.3.1
+module load fms/2023.02
+```
+The advantace is that it is obvious to the developer, the reviewers and the code managers that a library update in a PR is not officially rolled out everywhere yet. Further, there is no risk to corrupt official spack-stack installations with frequent test installs.
```
Comment (srherbener) [inline project_charter.md]:
```suggestion
The advantage is that it is obvious to the developer, the reviewers and the code managers that a library update in a PR is not officially rolled out everywhere yet. Further, there is no risk to corrupt official spack-stack installations with frequent test installs. 
```

File: project_charter.md [hunk 2]
```diff
@@ -63,28 +64,34 @@ execution, package files can be temporarily housed in the the spack
...
 
 ### Releases
 
 Any stakeholder can perform a release at any time.
 
 The release process:
-* Add a new issue explaining the reason for the release, and to notify all other stakeholders.
-* Designate any issues that must be fixed before release, and ensure their completion.
+* Add a new epic (github issue) explaining the reason for the release, and to notify all other stakeholders.
+* Unless it is a patch release (increasing the z in x.y.z,) create a project board for the release at https://github.com/JCSDA/spack-stack/actions.
```
Comment (srherbener) [inline project_charter.md]:
```suggestion
* Unless it is a patch release (increasing the z in x.y.z,) create a project board for the release at https://github.com/JCSDA/spack-stack/projects.
```

**Review by srherbener — COMMENTED** [review]
I think the updates look great, just found a couple typos.

File: project_charter.md [hunk 3]
```diff
@@ -103,3 +110,85 @@ Problems in spack package files will be solved by programmers of the
...
+#### Also to consider
+Our spack environments are large and complex. Adding more and more packages to the same environment between releases increases the risk of spack concretization errors (rebuilding existing libraries against new dependencies, unintended updates to existing libraries, duplicate packages, …)
+
+Spack-stack unified environments serve multiple applications at the same time: `ufs-weather-model`, `ufs-srw-app`, `skylab` (`jedi-fv3`, `soca`, `ewok`), `jedi-mpas`, `jedi-neptune`, `jedi-um`, `jedi-ufs`, etc. - for each of these, developers may potentially request several package installations “between releases”.
+
+### Handling library updates for applications
+We need to consider two types of installations: test installations and official updates
+
+#### Test installations
+Test installations can be requested as part of issues (which should be mandatory for pull requests anyway). This can be more frequent than official updates.
```
Comment (ulmononian) [inline project_charter.md]:
```suggestion
Test installations can be requested as part of issues (which should be mandatory for pull requests anyway). These can be more frequent than official updates.
```

File: project_charter.md [hunk 4]
```diff
@@ -103,3 +110,85 @@ Problems in spack package files will be solved by programmers of the
...
+module load fms/2023.02
+```
+The advantace is that it is obvious to the developer, the reviewers and the code managers that a library update in a PR is not officially rolled out everywhere yet. Further, there is no risk to corrupt official spack-stack installations with frequent test installs. 
+
+#### Official updates / patch releases
+Official updates should be limited in frequency and are only allowed after full acceptance testing (see test installations above).
+
+Updates that will generate module duplicates require a completely new spack-stack installation. Making these kinds of updates between releases should be the absolute exception when it is really urgent (critical bugs or security issues). We roll out a new spack-stack release every three months anyway!
+
+Official updates need to be made consistently across all platforms. There shall never be a situation when an official spack-stack installation on one platform differs from the others in their packages, package configurations etc (unless required by and encoded in spack-stack itself).
```
Comment (ulmononian) [inline project_charter.md]:
```suggestion
Official updates need to be made consistently across all platforms. There shall never be a situation when an official spack-stack installation on one platform differs from the others in their packages, package configurations, etc. (unless required by and encoded in spack-stack itself).
```

File: project_charter.md [hunk 5]
```diff
@@ -103,3 +110,85 @@ Problems in spack package files will be solved by programmers of the
...
+
+Since official updates require coordination across all platforms, the timeline for these installs must be more generous than for test installs.
+
+## Lifetime of spack-stack installations
+Regular, full releases of spack-stack are made every three months around end of February, May, August, November.
```
Comment (ulmononian) [inline project_charter.md]:
```suggestion
Regular, full releases of spack-stack are made every three months around the end of February, May, August, and November.
```

File: project_charter.md [hunk 6]
```diff
@@ -103,3 +110,85 @@ Problems in spack package files will be solved by programmers of the
...
+
+spack-stack developers are committed to maintaining a rolling stock of the last four releases, i.e. one year worth of spack-stack releases. Under special circumstances (e.g. operational implementations that require long-term support, long-running official retrospectives), selected releases may be kept longer.
```
Comment (ulmononian) [inline project_charter.md]:
```suggestion
spack-stack developers are committed to maintaining a rolling stock of the last four releases, i.e. one year's worth of spack-stack releases. Under special circumstances (e.g. operational implementations that require long-term support, long-running official retrospectives), selected releases may be kept longer.
```

**Review by ulmononian — CHANGES_REQUESTED** [review]

File: project_charter.md [hunk 7]
```diff
@@ -63,28 +64,34 @@ execution, package files can be temporarily housed in the the spack
...
+are necessary to keep the code in sync and simplify the process of
+sending updates in our fork to upstream.
+
 ### Installing on HPC Systems
 
 The spack-stack stack must be installed on various NOAA, NCAR, NASA,
-and other HPC systems. Each stakeholder will designate programmers to
-install on their own systems.
+and other HPC systems. Each stakeholder will designate developers to
+install on their systems (see https://spack-stack.readthedocs.io/en/latest/#).
```
Comment (AlexanderRichert-NOAA) [inline project_charter.md]: I assume this link should be https://spack-stack.readthedocs.io/en/latest/PreConfiguredSites.html

File: project_charter.md [hunk 8]
```diff
@@ -63,28 +64,34 @@ execution, package files can be temporarily housed in the the spack
...
 
 ### Releases
 
 Any stakeholder can perform a release at any time.
 
 The release process:
-* Add a new issue explaining the reason for the release, and to notify all other stakeholders.
-* Designate any issues that must be fixed before release, and ensure their completion.
+* Add a new epic (github issue) explaining the reason for the release, and to notify all other stakeholders.
```
Comment (AlexanderRichert-NOAA) [inline project_charter.md]: ->"Add a new EPIC GitHub issue"?

**Review by srherbener — APPROVED** [review]

**Review by ulmononian — APPROVED** [review]

**Pull request merged**